File: script/meta/minimyth/files/source/rootfs/usr/bin/mailnotifier.pl
#!/usr/bin/perl -w
#
# mailnotifier.pl - Declan Higgins, Piotr Oniszczuk
#
# Version 3.0 Sep 2013
#
# Scrip for every check_period checks for new mails 
# and if new are detected
# script sends notification to ip_list.
#
# Script should be called with parameter <check_period>
#
# Format of mailnotifier.rc file:
#
#    account,popserver,mbox1,password,limit,ip_list
#    account,popserver,mbox2,password,limit,ip_list
#    account,popserver,mbox3,password,limit,ip_list
#
# Legend:
# -"account" is a literal - dont edit it ...
# -"popserver" is pop server host name
# -"mbox[x]" is mailbox <UserID>
# -"password" is mailbox <password>
# -"limit" numeric field is limit the
#  number of emails notified (useful if
#  your mailbox has 500 emails). Not implemented due
#  limitation of current MythUiNotify
# -"ip_list" is list of notiefied FE IP addreses separated by ";"
#  i.e. 192.168.1.1;192.168.1.2
# Example:
#  account,pop3.o2.pl,johndoe,hispassword,5,127.0.0.1;192.168.1.1


use IO::Socket;
use Net::POP3;
use Net::Telnet;


######################################################################
#########   User Defined Variables                           #########
######################################################################

# Location of .rc file
my $config="/etc/mailnotifier.rc";

# Want to debug ?
my $debug=0;
my $debug2=0;

# Amount of sec for checking myth location and pooling POP accounts
my $check_period = $ARGV[0];

# How long osd dialog is displayed (if FE in in playback) 
my $osd_timeout=8;

# UI definition
my $new_mail_icon = "images/mythnotify/newmail.png";

our $new_mail_str;
our $accout_str;
our $mails_str;
#my $new_mail_str = "New E-Mail...";
#my $accout_str = "Account: ";
#my $mails_str = " new messages...";
require '/etc/mm_ui_localizations_perl';

# Whre sent notify
my $ip_list = "127.0.0.1";

######################################################################
#########   End User Defined Variables--Modify nothing below #########
######################################################################


my($ver)='3.0';
#my $pmc{}=-1;
my $mc=-1;
my $nmc= -1;
my $fake_out = 0;
my @notify_list = ();
my $new_mail = 0;

print "\n\n#############################\n";
print "Mail Notifier v. $ver\n";
print "Developed by: Declan Higgins\n";
print "Extended  by: Piotr Oniszczuk\n";
print "#############################\n\n";



sub stack_notify {
    my ($title,$origin,$description,$extra,$image,$progress_text,$progress,$timeout,$style,$ip_list) = @_;
    my @dest_list = ();
    my $msg = "";
    @dest_list = split(/;/, $ip_list);
    print ("OSD_notify:\n  title=\"$title\"\n  origin=\"$origin\"\n  description=\"$description\"\n  extra=\"$extra\"\n  image=\"$image\"\n  progress_txt=\"$progress_text\"\n  progress=$progress\n  style=\"$style\"\n") if ($debug);

    for (@dest_list) {
        print ("OSD_notify: notify OSD at IP=$_ \n") if ($debug2);
        $msg ="<mythnotification version=\"1\">
            <image>$image</image>
            <text>$title</text>
            <origin>$origin</origin>
            <description>$description</description>
            <extra>$extra</extra>
            <progress_text>$progress_text</progress_text>
            <progress>$progress</progress>
            <timeout>$timeout</timeout>";
        if ($style) { $msg = $msg."
            <type>$style</type>" } $msg = $msg."
        </mythnotification>";

        @notify_list = (@notify_list,$msg);

    }
}

sub do_account  () {
    my $unwanted = shift ;
    my $mail_server = shift ;
    my $username = shift ;
    my $password = shift ;
    my $maxmess = shift ;
    my $dest_ip = shift ;
    my $messno = 0;
    my $pmc;

    print("  Mail Srv is: $mail_server\n") if ($debug);
    print("  Username is: $username\n") if ($debug);
    print("  Password is: $password\n") if ($debug);

    $pop = Net::POP3->new($mail_server)
        or return $nmc;

    defined ($pop->login($username, $password))
        or return $nmc;

    $messages = $pop->list
        or return $nmc;

    $mc = keys %$messages;
    print("  Msg in $username: $mc\n") if ($debug);
    if ($pmc{$username} == $mc) {
        $pmc{$username} = $mc;
        print("  No new or deleted mails\n ") if ($debug);
        return $nmc;
    }

    print("  New or deleted mails detected. Was:$pmc{$username}, now is:$mc") if ($debug);
    $pmc{$username} = $mc;
    if ($mc==0) {
        return $nmc;
    }

    $nmc="[$username\@$mail_server]: $mc nowych wiad.";
    $new_mail = 1;

    stack_notify(
    #        title
    #        origin
    #        description
    #        extra
    #        image
    #        progress_text
    #        progress
    #        timeout
    #        style
    #        ip_list
    $new_mail_str,
    "", 
    $accout_str.$username."@".$mail_server,
    $mc.$mails_str,
    $new_mail_icon,
    "",
    "",
    $osd_timeout,
    "",
    $ip_list
    );

    return $nmc;
}

sleep (120);

while (1) {

    open CONFIG,"<$config";

    while (<CONFIG>) {
        print("\nChecking account\n") if ($debug);
        chomp;
        @b=split /,/;
        next unless $b[0] eq "account";
            my $msg = &do_account (@b);
    }

    if ($new_mail) {
        my $item;
        my @dest_list = ();
        @dest_list = split(/,/, $ip_list);

        for (@dest_list) {
            foreach $item (@notify_list) {
                my $mythnotify_fh = IO::Socket::INET->new(PeerAddr=>$_,Proto=>'udp',PeerPort=>6948);
                if ($mythnotify_fh) {
                    print $mythnotify_fh $item;
                    $mythnotify_fh->close;
                    print ("Notify via OSD to IP=$_ done\n") if ($debug);
                }
            }
        }

        $new_mail = 0;
        @notify_list = ();
    }

    print("Sleeping $check_period sec.\n") if ($debug);

    sleep ($check_period);

}
1

File: script/meta/minimyth/files/source/rootfs/usr/bin/mailnotifier.py
#!/usr/bin/python -W ignore::DeprecationWarning
# -*- coding: UTF-8 -*-

import sys
import time
import re
import poplib
import socket
from imapclient import IMAPClient
from collections import defaultdict

debug = 1
cfg_file = "/etc/mailnotifier.rc"









if debug:
    print("\n\nMail notifier v1.1 (c) Piotr Oniszczuk\n\n")

argc = len(sys.argv)

if argc < 2:
    sleep = "60"
    if debug:
        print("Check period is:" + sleep + "sec.\n")
else:
    sleep = sys.argv[1]
    if debug:
        print("Check period is:" + sleep + "sec.\n\nInitial delaying of check by " + sleep + "sec.\n")
    time.sleep(int(sleep))

current_mails = defaultdict(int)

def send_osd (IP, TITLE, ORIGIN, DESCRIPTION, EXTRA, IMAGE, PROGRESS_TEXT, PROGRESS, TIMEOUT, STYLE):

    PORT = 6948

    MESSAGE = "\n<mythnotification version=\"1\">\n" + \
        "<image>"         + IMAGE         + "</image>\n" + \
        "<text>"          + TITLE         + "</text>\n" + \
        "<origin>"        + ORIGIN        + "</origin>\n" + \
        "<description>"   + DESCRIPTION   + "</description>\n" + \
        "<extra>"         + EXTRA         + "</extra>\n" + \
        "<progress_text>" + PROGRESS_TEXT + "</progress_text>\n" + \
        "<progress>"      + PROGRESS      + "</progress>\n" + \
        "<timeout>"       + TIMEOUT       + "</timeout>\n"
    if STYLE:
        MESSAGE = MESSAGE + "<style>" + STYLE + "</style>\n"

    MESSAGE = MESSAGE + "</mythnotification>"

    if debug:
        print("OSD_notify:")
        print("Target IP:", IP)
        print("Target port:", PORT)
        print("Message:", MESSAGE)

    sock = socket.socket(socket.AF_INET,socket.SOCK_DGRAM)
    sock.sendto(bytes(MESSAGE, 'utf-8'), (IP, PORT))




while True:
    file = open(cfg_file,"r")
    lines = filter(None, (line.rstrip() for line in file))
    for line in lines:
        line = re.sub('\s*', "", line)
        if not re.search('^#|^;', line):
            line = re.sub('\n', "", line)
            x = line.split(",")
            type     = x[0]
            server   = x[1]
            login    = x[2]
            password = x[3]
            timeout  = x[4]
            ip       = x[5]
            picture  = x[6]

            if debug:
                print("Checking account:" + \
                "\n  type     : " + type + \
                "\n  server   : " + server + \
                "\n  login    : " + login + \
                "\n  password : " + password + \
                "\n  timeout  : " + timeout + \
                "\n  ip       : " + ip + \
                "\n  picture  : " + picture)

            cleanlogin = re.sub('\@', "-at-", login)

            if type == "imap":
                mailbox = "Inbox"
                imapserver = IMAPClient(server, use_uid=True, ssl=True)
                try:
                    imapserver.login(login, password)

                except:
                    print('Cant login to ' + login)
                    send_osd(ip, "Konto E-Mail:", "", "Nie mogę się zalogować do: " + login, "", picture, "", "", timeout, "")

                else:
                    if debug:
                        print('Logging in as ' + login)

                    select_info = imapserver.select_folder(mailbox)

                    if debug:
                        print('%d messages in INBOX' % select_info[b'EXISTS'])

                    folder_status = imapserver.folder_status(mailbox, b'UNSEEN')
                    newmails = int(folder_status[b'UNSEEN'])

                    if debug:
                        print("Account have", newmails, "new emails!")

                    if newmails > current_mails[cleanlogin]:
                        current_mails[cleanlogin] = newmails
                        if debug:
                            print("Sending notification about new emails!")
                        send_osd(ip, "Konto E-Mail:", "", str(newmails) + " nowych wiad.", picture, "", "", timeout, "", "")
                    else:
                        current_mails[cleanlogin] = newmails
                        if debug:
                            print("No notification about new emails!")

            if (type == "pop3" or type == "pop3ssl"):
                if type == "pop3ssl":
                    pop3server = poplib.POP3(server)
                else:
                    pop3server = poplib.POP3_SSL(server)

                try:
                    pop3server.user(login)
                    pop3server.pass_(password)
                    stat = pop3server.stat()

                except:
                    print('Cant login to ' + login)
                    send_osd(ip, "Konto E-Mail:", "", "Nie mogę się zalogować do: " + login, "", picture, "", "", timeout, "")

                else:
                    newmails = stat[0]

                    if debug:
                        print("Account have", newmails, "new emails!")

                    if newmails > current_mails[cleanlogin]:
                        if debug:
                            print("Sending notification about new emails!")
                        current_mails[cleanlogin] = newmails
                        send_osd(ip, "Konto E-Mail:", "", login + "@" + server, str(newmails) + " nowych wiad.", picture, "", "", timeout, "")
                    else:
                        current_mails[cleanlogin] = newmails
                        if debug:
                            print("No notification about new emails!")


    if debug:
        print(current_mails)

    time.sleep(int(sleep))

